### PR TITLE
ISPN-14618 ZRANGE

### PR DESCRIPTION
**Pull request by karesti**
https://issues.redhat.com/browse/ISPN-14618

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 1]
```diff
@@ -3,49 +3,262 @@
...
+ * @since 15.0
  */
 public class ZRANGE extends RespCommand implements Resp3Command {
+   protected enum Arg {
+      BYSCORE, BYLEX, REV, LIMIT, WITHSCORES
+   }
+
+   private boolean isRev;
+   private boolean byScore;
+   private boolean byLex;
```
Comment (jabolina) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:48]: I think using these variables as instance variables might be causing an issue. The ZRANGE is shared between all client connections, so we need to synchronize the access. And we are not resetting after execution, so mixing the flags raise issues:

Here is one client in one terminal:

```text
127.0.0.1:11222> zrange my-set 0 3 BYSCORE
1) "a"
2) "b"
3) "c"
127.0.0.1:11222> zrange my-set a c BYLEX
(error) ERR syntax error
127.0.0.1:11222> zrange my-set 0 3 BYSCORE
(error) ERR syntax error
```

And another client in another terminal, freshly connected:

```text
127.0.0.1:11222> auth admin password
OK
127.0.0.1:11222> zrange my-set 0 3
(error) ERR syntax error
```

**Review by jabolina — CHANGES_REQUESTED** [review]
The first pass, a few things, and I think we're good.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java (hunk 1 above)
Comment (karesti) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:48]: yes, I realized this and solved this in ZRANGESTORE

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 2]
```diff
@@ -3,49 +3,265 @@
...
-                  result.add(Double.toString(e.score()).getBytes(StandardCharsets.US_ASCII));
-               });
-               return result;
-            });
+      byte[] start = arguments.get(1);
+      byte[] stop = arguments.get(2);
+
+      int pos = 3;
+      final ResultOptions resultOpt = new ResultOptions();
+      boolean isRev = this.initialRev;
```
Comment (karesti) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:88]: I changed here to use local variables

File: multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java [hunk 3]
```diff
@@ -144,55 +143,180 @@ private void addScoredValue(Double newScore, MultimapObjectWrapper<V> value) {
...
+   }
+
    @SuppressWarnings({ "unchecked", "rawtypes" })
-   public SortedSet<ScoredValue<V>> subsetByScores(double min, boolean includeMin, double max, boolean includeMax) {
-      ScoredValue<V> minSv;
-      ScoredValue<V> maxSv;
-      boolean unboundedMin = false;
-      boolean unboundedMax = false;
-      if (min == max && (!includeMin || !includeMax)) {
+   public SortedSet<ScoredValue<V>> subset(Double startScore, boolean includeStart, Double stopScore, boolean includeStop, boolean isRev) {
```
[the remark this answers is not in the document]
Comment (jabolina) [reply multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java:185]: The methods in the collections are also returning the backing collection, the `descendingSet`, `subSet`, etc.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 4]
```diff
@@ -3,49 +3,265 @@
...
+                                                                      EmbeddedMultimapSortedSetCache<byte[], byte[]> sortedSetCache) {
+      // Size is called to raise ERR type if needed
+      return handler.stageToReturn(sortedSetCache.size(name)
+                  .thenApply(r -> Collections.emptyList()), ctx, Consumers.GET_ARRAY_BICONSUMER);
+   }
+
+   /**
+    * Transforms the resulting collection depending on the zrank options
+    * - return scores
+    * - limit results
```
Comment (jabolina) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:232]: Thinking if we send these parameters to limit the remote results. Maybe we add it later.

File: server/resp/src/test/java/org/infinispan/server/resp/SortedSetCommandsTest.java [hunk 5]
```diff
@@ -244,15 +246,510 @@ public void testZPOPMAX() {
...
+            just(18.9, "fabio"),
+            just(18.9, "jose"),
+            just(18.9, "katia"),
+            just(21.9, "marc"));
       assertThat(redis.zscore("people", "jose")).isEqualTo(18.9);
       assertThat(redis.zscore("people", "takolo")).isNull();
       assertWrongType(() -> redis.set("another", "tristan"), () ->  redis.zscore("another", "tristan"));
    }
+
+   public void testZRANGE() {
```
Comment (jabolina) [inline server/resp/src/test/java/org/infinispan/server/resp/SortedSetCommandsTest.java:261]: More of a nitpick. We really cover each one separately with multiple combinations. Can we add just a simple one that mixes the commands?

File: multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (karesti) [reply multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java:185]: those are ok, it's a copy

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java (hunk 4 above)
Comment (karesti) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:232]: yes, I've thought on that too as an improvement of perf, we need to run some tests to go into improv

File: server/resp/src/test/java/org/infinispan/server/resp/SortedSetCommandsTest.java (hunk 5 above)
Comment (karesti) [reply server/resp/src/test/java/org/infinispan/server/resp/SortedSetCommandsTest.java:261]: methods are already too big already. lettuce does not support all the options in zrange, but keeps the old commands. the code is the same, with the same parameters sent.

**Review by jabolina — APPROVED** [review]

**Comment by jabolina** [comment]
Integrated, thanks @karesti!

**Pull request merged**